lib/Analysis/NoiseAnalysis/BGV/NoiseByVarianceCanEmbModel.cpp

```diff
@@ -0,0 +1,300 @@
+#include "lib/Analysis/NoiseAnalysis/BGV/NoiseByVarianceCanEmbModel.h"
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <iomanip>
+#include <ios>
+#include <numeric>
+#include <sstream>
+#include <string>
+
+namespace mlir {
+namespace heir {
+namespace bgv {
+
+// the formulae below are mainly taken from MMLGA22
+// "Finding and Evaluating Parameters for BGV"
+// https://eprint.iacr.org/2022/706
+
+template <bool P>
+using Model = NoiseByVarianceCanEmbModel<P>;
+
+template <bool P>
+double Model<P>::toLogBound(const LocalParamType &param,
+                            const StateType &noise) {
+  auto t = param.getSchemeParam()->getPlaintextModulus();
+  // StateType only stores e in (m + t * e), so when we want to print the bound
+  // we need to multiply t back
+  return log(t * noise.getValue()) / log(2);
```

Comment:
From `evalEncryptPk` below it seems that `StateType` stores `||m + te||` as you multiply `t` to the `D\sqrt{variance-stuff}`. Then here we do not need to multiply `t` again.

This sentence was here because KPZ21 only tracks `||e||` instead of `||m + te||`. For MMLGA22 it seems that it is not the case.

Reply:
Fixed it